sandboxing breaks ccache

I'm now using opam2~rc2. On my Fedora system, running

```
opam switch create . 4.06.1
```

fails with the following error

```
#=== ERROR while compiling ocaml-base-compiler.4.06.1 =========================#
# context     2.0.0~rc | linux/x86_64 |  | https://opam.ocaml.org/2.0#bed714e6
# path        ~/tmp/repro/_opam/.opam-switch/build/ocaml-base-compiler.4.06.1
# command     ~/.opam/opam-init/hooks/sandbox.sh build make world
# exit-code   2
# env-file    ~/.opam/log/ocaml-base-compiler-22947-1872ec.env
# output-file ~/.opam/log/ocaml-base-compiler-22947-1872ec.out
### output ###
# make -C byterun  all
# make[1]: Entering directory '/home/gasche/tmp/repro/_opam/.opam-switch/build/ocaml-base-compiler.4.06.1/byterun'
# echo "/home/gasche/tmp/repro/_opam/lib/ocaml/stublibs" > ld.conf
# echo "/home/gasche/tmp/repro/_opam/lib/ocaml" >> ld.conf
# sed -n -e '/^  /s/ \([A-Z]\)/ \&\&lbl_\1/gp' \
#        -e '/^}/q' caml/instruct.h > caml/jumptbl.h
# gcc -c -O2 -fno-strict-aliasing -fwrapv -Wall -fno-tree-vrp  -g -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DCAML_NAME_SPACE  interp.c
# ccache: error: Failed to create temporary file for /home/gasche/.ccache/tmp/tmp.cpp_stderr: Read-only file system
# make[1]: *** [Makefile:205: interp.o] Error 1
# make[1]: Leaving directory '/home/gasche/tmp/repro/_opam/.opam-switch/build/ocaml-base-compiler.4.06.1/byterun'
# make: *** [Makefile:384: coldstart] Error 2
```

It turns out (I had no idea) that `ccache` is enabled by default on Fedora OSes. Could you add `$HOME/.ccache` to the list of directories to the write whitelist of the sandbox?

For reference (if other users encounter the issue), a local fix is to pass `CCACHE_DISABLE=1` explicitly to the command-line:

```
CCACHE_DISABLE=1 opam switch create . 4.06.1
```

of course, many other OPAM packages (than the compiler) invoke `gcc`, so selectively disabling ccache all the time will get old fast -- then export it in your shell environment, etc.

--- a/src/state/shellscripts/bwrap.sh
+++ b/src/state/shellscripts/bwrap.sh
@@ -25,12 +25,29 @@ add_mounts() {
 
 add_mounts ro /usr /bin /lib /lib32 /lib64 /etc /opt /nix/store /home
 
+# C compilers using `ccache` will write to a shared cache directory
+# that remain writeable. ccache seems widespread in some Fedora systems.
+add_ccache_mount() {
+  if command -v ccache > /dev/null; then
+      CCACHE_DIR=$HOME/.ccache
+      ccache_dir_regex='cache_dir = (.*)$'
+      local IFS=$'\n'
+      for f in $(ccache --print-config); do
+        if [[ $f =~ $ccache_dir_regex ]]; then
+          CCACHE_DIR=${BASH_REMATCH[1]}
+        fi
+      done
+      add_mounts rw $CCACHE_DIR
+  fi
+}
+
 # This case-switch should remain identical between the different sandbox implems
 COMMAND="$1"; shift
 case "$COMMAND" in
     build)
         add_mounts ro "$OPAM_SWITCH_PREFIX"
         add_mounts rw "$PWD"
+        add_ccache_mount
         ;;
     install)
         add_mounts rw "$OPAM_SWITCH_PREFIX"

--- a/src/state/shellscripts/sandbox_exec.sh
+++ b/src/state/shellscripts/sandbox_exec.sh
@@ -1,4 +1,4 @@
-#!/bin/sh -ue
+#!/bin/bash -ue
 
 POL='(version 1)(allow default)(deny network*)(deny file-write*)'
 POL="$POL"'(allow file-write* (literal "/dev/null"))'

--- a/src/state/shellscripts/sandbox_exec.sh
+++ b/src/state/shellscripts/sandbox_exec.sh
@@ -13,12 +13,29 @@ add_mounts() {
 
 add_mounts rw "${TMPDIR:-/tmp}"
 
+# C compilers using `ccache` will write to a shared cache directory
+# that remain writeable. ccache seems widespread in some Fedora systems.
+add_ccache_mount() {
+  if command -v ccache > /dev/null; then
+      CCACHE_DIR=$HOME/.ccache
+      ccache_dir_regex='cache_dir = (.*)$'
+      local IFS=$'\n'
+      for f in $(ccache --print-config); do
+        if [[ $f =~ $ccache_dir_regex ]]; then
+          CCACHE_DIR=${BASH_REMATCH[1]}
+        fi
+      done
+      add_mounts rw $CCACHE_DIR
+  fi
+}
+
 # This case-switch should remain identical between the different sandbox implems
 COMMAND="$1"; shift
 case "$COMMAND" in
     build)
         add_mounts ro "$OPAM_SWITCH_PREFIX"
         add_mounts rw "$PWD"
+        add_ccache_mount
         ;;
     install)
         add_mounts rw "$OPAM_SWITCH_PREFIX"

